exporter.go

```diff
@@ -63,75 +66,133 @@ func hashNameAndLabels(name string, labels prometheus.Labels) uint64 {
 	return hash.Sum64()
 }
 
+func labelsNames(labels prometheus.Labels) []string {
+	names := make([]string, 0, len(labels))
+	for labelName := range labels {
+		names = append(names, labelName)
+	}
+	sort.Strings(names)
+	return names
+}
+
+// hashNameAndLabels returns a hash value of the provided name string and all
+// the label names in the provided labels map.
+//
+// Not safe for concurrent use! (Uses a shared buffer and hasher to save on
+// allocations.)
+func hashNameAndLabelNames(name string, labels prometheus.Labels) uint64 {
+	hash.Reset()
+	strBuf.Reset()
+	strBuf.WriteString(name)
+	for _, labelName := range labelsNames(labels) {
+		strBuf.WriteByte(255)
+		strBuf.WriteString(labelName)
+	}
+	hash.Write(strBuf.Bytes())
+	return hash.Sum64()
+}
+
 type CounterContainer struct {
-	Elements map[uint64]prometheus.Counter
+	//               metric     labels names hash
+	Elements     map[string]map[uint64]*prometheus.CounterVec
 }
 
 func NewCounterContainer() *CounterContainer {
 	return &CounterContainer{
-		Elements: make(map[uint64]prometheus.Counter),
+		Elements:     make(map[string]map[uint64]*prometheus.CounterVec),
 	}
 }
 
 func (c *CounterContainer) Get(metricName string, labels prometheus.Labels, help string) (prometheus.Counter, error) {
-	hash := hashNameAndLabels(metricName, labels)
-	counter, ok := c.Elements[hash]
+	hash := hashNameAndLabelNames(metricName, labels)
+	var counter *prometheus.CounterVec
+	_, ok := c.Elements[metricName]
 	if !ok {
-		counter = prometheus.NewCounter(prometheus.CounterOpts{
-			Name:        metricName,
-			Help:        help,
-			ConstLabels: labels,
-		})
+		counter = prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: metricName,
+			Help: help,
+		}, labelsNames(labels))
 		if err := prometheus.Register(counter); err != nil {
 			return nil, err
 		}
-		c.Elements[hash] = counter
+		c.Elements[metricName] = map[uint64]*prometheus.CounterVec{
+			hash: counter,
+		}
+	}
+	counter, ok = c.Elements[metricName][hash]
+	if !ok {
+		err := fmt.Errorf(regErrNames, labelsNames(labels))
+		return nil, err
+	}
+
+	return counter.With(labels), nil
+}
+
+func (c *CounterContainer) Delete(metricName string, labels prometheus.Labels) {
+	if _, ok := c.Elements[metricName]; ok {
+		hash := hashNameAndLabelNames(metricName, labels)
+		c.Elements[metricName][hash].Delete(labels)
 	}
-	return counter, nil
 }
 
 type GaugeContainer struct {
-	Elements map[uint64]prometheus.Gauge
+	Elements map[string]map[uint64]*prometheus.GaugeVec
 }
 
 func NewGaugeContainer() *GaugeContainer {
 	return &GaugeContainer{
-		Elements: make(map[uint64]prometheus.Gauge),
+		Elements: make(map[string]map[uint64]*prometheus.GaugeVec),
 	}
 }
 
 func (c *GaugeContainer) Get(metricName string, labels prometheus.Labels, help string) (prometheus.Gauge, error) {
-	hash := hashNameAndLabels(metricName, labels)
-	gauge, ok := c.Elements[hash]
+	hash := hashNameAndLabelNames(metricName, labels)
```

Comment:
how is this guarded against concurrent usage?

Reply:
Same as before: methods of Exporter and *Container structs are called synchronously. Each received statsd line is transformed into a set of events. Events are send to a channel in a single instance of Exporter and handled there one by one. 

https://github.com/prometheus/statsd_exporter/blob/master/main.go#L211-L212

https://github.com/flant/statsd_exporter/blob/remove_stale_metrics/exporter.go#L359

Reply:
Ah cool, that makes sense.